plugins/woocommerce/client/legacy/js/admin/meta-boxes-product.js

```diff
@@ -447,59 +455,84 @@ jQuery( function ( $ ) {
 		selectedAttributes
 	);
 
-	function add_attribute( element, attribute ) {
-		var size = $( '.product_attributes .woocommerce_attribute' ).length;
-		var $wrapper = $( element ).closest( '#product_attributes' );
-		var $attributes = $wrapper.find( '.product_attributes' );
-		var product_type = $( 'select#product-type' ).val();
-		var data = {
-			action: 'woocommerce_add_attribute',
-			taxonomy: attribute,
-			i: size,
-			security: woocommerce_admin_meta_boxes.add_attribute_nonce,
-		};
+	function get_new_attribute_list_item_html( indexInList, globalAttributeId ) {
+		return new Promise( function ( resolve, reject ) {
+			$.post( {
+				url: woocommerce_admin_meta_boxes.ajax_url,
+				data: {
+					action: 'woocommerce_add_attribute',
+					taxonomy: globalAttributeId ?? '',
+					i: indexInList,
+					security: woocommerce_admin_meta_boxes.add_attribute_nonce,
+				},
+				success: function ( newAttributeListItemHtml ) {
+					resolve( newAttributeListItemHtml );
+				},
+				error: function ( jqXHR, textStatus, errorThrown ) {
+					reject( { jqXHR, textStatus, errorThrown } );
+				}
+			} );
+		} );
+	}
+
+	function block_attributes_tab_container() {
+		const $attributesTabContainer = $( '#product_attributes' );
 
-		$wrapper.block( {
+		$attributesTabContainer.block( {
 			message: null,
 			overlayCSS: {
 				background: '#fff',
 				opacity: 0.6,
 			},
 		} );
+	}
 
-		$.post( woocommerce_admin_meta_boxes.ajax_url, data, function (
-			response
-		) {
-			$attributes.append( response );
+	function unblock_attributes_tab_container() {
+		const $attributesTabContainer = $( '#product_attributes' );
+		$attributesTabContainer.unblock();
+	}
 
-			if ( 'variable' !== product_type ) {
-				$attributes.find( '.enable_variation' ).hide();
-			}
+	function toggle_expansion_of_attribute_list_item( $attributeListItem ) {
+		$attributeListItem.find( 'h3' ).trigger( 'click' );
+	}
 
-			$( document.body ).trigger( 'wc-enhanced-select-init' );
+	function init_select_controls() {
+		$( document.body ).trigger( 'wc-enhanced-select-init' );
+	}
 
-			attribute_row_indexes();
+	async function add_attribute_to_list( globalAttributeId ) {
+		try {
```

Comment:
I think that if we use exception handling, we should add a `catch` to this `try`.

Reply:
> I could put up an alert... do you think that is good enough?

An alert mentioning the error sounds good to me 👍

Reply:
Sounds good. I'll make that change and fix the conflicting test file right after lunch and ping you when ready for another review.